File: src/js/multithreading/09_merge_sort.js
/*
 * Parallel Merge Sort with Worker Threads
 *
 * This script demonstrates parallelizing the merge sort algorithm using
 * worker threads. It divides the array into chunks, sorts each chunk
 * in a separate worker, then merges the results.
 *
 * Key concepts:
 * - Divide and conquer parallelization
 * - Worker thread pool for parallel chunk sorting
 * - Performance comparison: single-threaded vs multi-threaded
 */

"use strict";

const { Worker, isMainThread, parentPort } = require("worker_threads");
const os = require("os");

const NUM_CPUS = os.cpus().length;

const merge = (left, right) => {
  const result = new Array(left.length + right.length);
  let leftIdx = 0;
  let rightIdx = 0;
  let resultIdx = 0;

  while (leftIdx < left.length && rightIdx < right.length) {
    if (left[leftIdx] < right[rightIdx]) {
      result[resultIdx++] = left[leftIdx++];
    } else {
      result[resultIdx++] = right[rightIdx++];
    }
  }

  while (leftIdx < left.length) {
    result[resultIdx++] = left[leftIdx++];
  }
  
  while (rightIdx < right.length) {
    result[resultIdx++] = right[rightIdx++];
  }

  return result;
};

const mergeSort = (arr) => {
  if (arr.length <= 1) {
    return arr;
  }

  const mid = Math.floor(arr.length / 2);
  const left = mergeSort(arr.slice(0, mid));
  const right = mergeSort(arr.slice(mid));

  return merge(left, right);
};

const parallelMergeSort = async (arr) => {
  const chunkSize = Math.ceil(arr.length / NUM_CPUS);
  const chunks = [];

  // Split array into chunks
  for (let i = 0; i < arr.length; i += chunkSize) {
    chunks.push(arr.slice(i, i + chunkSize));
  }

  // Sort each chunk in parallel using workers
  const sortPromises = chunks.map((chunk) => {
    return new Promise((resolve, reject) => {
      const worker = new Worker(__filename);
      
      worker.postMessage(chunk);
      
      worker.on("message", (sortedChunk) => {
        resolve(sortedChunk);
      });
      
      worker.on("error", reject);
    });
  });

  const sortedChunks = await Promise.all(sortPromises);

  // Merge all sorted chunks
  return sortedChunks.reduce((acc, chunk) => merge(acc, chunk), []);
};

const generateRandomArray = (size) => {
  return Array.from({ length: size }, () => Math.floor(Math.random() * 100000));
};

const isSorted = (arr) => {
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] < arr[i - 1]) return false;
  }
  return true;
};

const main = async () => {
  console.log("=== Parallel Merge Sort Demo ===\n");
  console.log(`CPU cores: ${NUM_CPUS}`);

  const ARRAY_SIZE = 100000;
  const arr = generateRandomArray(ARRAY_SIZE);
  console.log(`Array size: ${ARRAY_SIZE.toLocaleString()} elements\n`);

  // Single-threaded merge sort
  console.log("--- Single-threaded Merge Sort ---");
  const singleStart = Date.now();
  const singleSorted = mergeSort([...arr]);
  const singleElapsed = Date.now() - singleStart;
  console.log(`Time: ${singleElapsed}ms`);
  console.log(`Correctly sorted: ${isSorted(singleSorted)}\n`);

  // Parallel merge sort
  console.log("--- Parallel Merge Sort ---");
  const parallelStart = Date.now();
  const parallelSorted = await parallelMergeSort([...arr]);
  const parallelElapsed = Date.now() - parallelStart;
  console.log(`Time: ${parallelElapsed}ms`);
  console.log(`Correctly sorted: ${isSorted(parallelSorted)}\n`);

  // Summary
  console.log("=== Performance Summary ===");
  console.log(`Single-threaded: ${singleElapsed}ms`);
  console.log(`Parallel:        ${parallelElapsed}ms`);
  
  if (parallelElapsed < singleElapsed) {
    const speedup = (singleElapsed / parallelElapsed).toFixed(2);
    console.log(`Speedup:         ${speedup}x`);
  } else {
    console.log("Note: Parallel may be slower for small arrays due to thread overhead");
  }
};

if (isMainThread) {
  main();
} else {
  parentPort.once("message", (arr) => {
    const sorted = mergeSort(arr);
    parentPort.postMessage(sorted);
  });
}

File: src/cpp/multithreading/09_merge_sort.cpp
#include <algorithm>
#include <chrono>
#include <iostream>
#include <iterator>
#include <random>
#include <thread>
#include <vector>

const int numThreads =
    10; // Change this value to control the number of threads used

void merge(std::vector<int> &arr, int left, int mid, int right) {
  int n1 = mid - left + 1;
  int n2 = right - mid;

  std::vector<int> L(n1);
  std::vector<int> R(n2);

  std::copy(arr.begin() + left, arr.begin() + mid + 1, L.begin());
  std::copy(arr.begin() + mid + 1, arr.begin() + right + 1, R.begin());

  int i = 0, j = 0, k = left;
  while (i < n1 && j < n2) {
    if (L[i] <= R[j]) {
      arr[k] = L[i];
      i++;
    } else {
      arr[k] = R[j];
      j++;
    }
    k++;
  }

  while (i < n1) {
    arr[k] = L[i];
    i++;
    k++;
  }

  while (j < n2) {
    arr[k] = R[j];
    j++;
    k++;
  }
}

void mergeSort(std::vector<int> &arr, int left, int right, int depth = 0) {
  if (left < right) {
    int mid = left + (right - left) / 2;
    if (depth < numThreads) {
      std::thread leftThread(mergeSort, std::ref(arr), left, mid, depth + 1);
      std::thread rightThread(mergeSort, std::ref(arr), mid + 1, right,
                              depth + 1);
      leftThread.join();
      rightThread.join();
    } else {
      mergeSort(arr, left, mid, depth + 1);
      mergeSort(arr, mid + 1, right, depth + 1);
    }
    merge(arr, left, mid, right);
  }
}

int main() {
  int n = 100000;
  std::vector<int> arr(n);
  std::iota(arr.begin(), arr.end(), 1);
  std::shuffle(arr.begin(), arr.end(), std::mt19937{std::random_device{}()});

  std::vector<int> arrCopy = arr;

  auto start = std::chrono::high_resolution_clock::now();
  mergeSort(arr, 0, arr.size() - 1);
  auto end = std::chrono::high_resolution_clock::now();
  std::chrono::duration<double> duration = end - start;

  std::cout << "Threaded merge sort took: " << duration.count() << " seconds"
            << std::endl;

  start = std::chrono::high_resolution_clock::now();
  std::sort(arrCopy.begin(), arrCopy.end());
  end = std::chrono::high_resolution_clock::now();
  duration = end - start;

  std::cout << "Non-threaded merge sort took: " << duration.count()
            << " seconds" << std::endl;

  return 0;
}
